File: imp-haskell/src/Eval.hs
module Eval where

import Ast
import Data.Map as M

type Store = M.Map String Int

evalAExp :: Store -> AExp -> Maybe Int
evalAExp _ (NUM val) = Just val
evalAExp st (LOC name) = M.lookup name st
evalAExp st (ADD e1 e2) = (+) <$> evalAExp st e1 <*> evalAExp st e2
evalAExp st (SUB e1 e2) = (-) <$> evalAExp st e1 <*> evalAExp st e2
evalAExp st (MUL e1 e2) = (*) <$> evalAExp st e1 <*> evalAExp st e2

evalBExp :: Store -> BExp -> Maybe Bool
evalBExp _ TRUE = Just True
evalBExp _ FALSE = Just False
evalBExp st (EQUAL e1 e2) = (==) <$> evalBExp st e1 <*> evalBExp st e2
evalBExp st (LEQ e1 e2) = (<=) <$> evalBExp st e1 <*> evalBExp st e2
evalBExp st (NOT e) = not <$> evalBExp st e
evalBExp st (AND e1 e2) = (&&) <$> evalBExp st e1 <*> evalBExp st e2
evalBExp st (OR e1 e2) = (||) <$> evalBExp st e1 <*> evalBExp st e2

evalCommand :: Store -> Command -> Maybe Store
evalCommand st SKIP = Just st
evalCommand st (ASSIGN name e) = (\v -> M.insert name v st) <$> evalAExp st e
evalCommand st (SEQ c1 c2) = do
  st' <- evalCommand st c1
  evalCommand st' c2
evalCommand st (IF p c a) = do
  pv <- evalBExp st p
  if pv
    then evalCommand st c
    else evalCommand st a
evalCommand st w@(WHILE p body) = do
  pv <- evalBExp st p
  if pv
    then evalCommand st (SEQ body w)
    else return st

File: imp-ocaml/eval.ml
open Ast

type store = (string * int) list

let rec eval_aexp (st, e) : int =
  match e with
    | Num n        -> n
    | Var name     -> List.assoc name st
    | Add (e1, e2) -> eval_aexp (st, e1) + eval_aexp (st, e2)
    | Sub (e1, e2) -> eval_aexp (st, e1) - eval_aexp (st, e2)
    | Mul (e1, e2) -> eval_aexp (st, e1) * eval_aexp (st, e2)

let rec eval_bexp (st, e) : bool =
  match e with
    | True           -> true
    | False          -> true
    | Equal (e1, e2) -> eval_bexp (st, e1) == eval_bexp (st, e2)
    | Leq (e1, e2)   -> eval_bexp (st, e1) <= eval_bexp (st, e2)
    | Not e          -> eval_bexp (st, e) |> not
    | And (e1, e2)   -> eval_bexp (st, e1) && eval_bexp (st, e2)
    | Or (e1, e2)    -> eval_bexp (st, e1) || eval_bexp (st, e2)

let rec eval_command (st, e) : store =
  match e with
    | Skip -> st
    | Assign (var, e) -> List.cons (var, eval_aexp (st, e)) st
    | Seq (c1, c2) ->
      let st' = eval_command (st, c1)
      in eval_command (st', c2)
    | If (p, c, a) ->
      if eval_bexp (st, p)
      then eval_command (st, c)
      else eval_command (st, a)
    | While (p, c) ->
      if eval_bexp (st, p)
      then eval_command (st, Seq(c, e))
      else st
